lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/TestS3Retries.java

```diff
@@ -0,0 +1,122 @@
+import java.util.Arrays;
+
+import static io.trino.testing.containers.Minio.MINIO_API_PORT;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+public class TestS3Retries
+{
+    Network network;
+    private Minio minio;
+    private ToxiproxyContainer toxiproxy;
+
+    @BeforeAll
+    final void init()
+            throws IOException
+    {
+        Logging logging = Logging.initialize();
+        logging.setLevel("software.amazon.awssdk.request", Level.DEBUG);
+
+        network = Network.newNetwork();
+        minio = Minio.builder()
+                .withNetwork(network)
+                .build();
+        minio.start();
+        minio.createBucket("bucket");
+        minio.writeFile(getTestData(), "bucket", "object");
+
+        toxiproxy = new ToxiproxyContainer("ghcr.io/shopify/toxiproxy:2.5.0")
+                .withExposedPorts(8474, 1234)
+                .withNetwork(network)
+                .withNetworkAliases("minio");
+        toxiproxy.start();
+
+        ToxiproxyClient toxiproxyClient = new ToxiproxyClient(toxiproxy.getHost(), toxiproxy.getControlPort());
+        Proxy proxy = toxiproxyClient.createProxy("minio", "0.0.0.0:1234", "minio:" + MINIO_API_PORT);
+        // the number of bytes includes both the response headers (around 570 bytes) and body
+        proxy.toxics()
+                .limitData("broken connection", ToxicDirection.DOWNSTREAM, 700);
+    }
+
+    private static byte[] getTestData()
+    {
+        byte[] data = new byte[1024];
+        Arrays.fill(data, (byte) 1);
+        return data;
+    }
+
+    @AfterAll
+    final void cleanup()
+    {
+        minio.close();
+        toxiproxy.close();
+        network.close();
+    }
+
+    private S3Client createS3Client()
+    {
+        return S3Client.builder()
+                .endpointOverride(URI.create("http://" + toxiproxy.getHost() + ":" + toxiproxy.getMappedPort(1234)))
+                .region(Region.of(Minio.MINIO_REGION))
+                .forcePathStyle(true)
+                .credentialsProvider(StaticCredentialsProvider.create(
+                        AwsBasicCredentials.create(Minio.MINIO_ACCESS_KEY, Minio.MINIO_SECRET_KEY)))
+                // explicitly configure the number of retries
+                .overrideConfiguration(o -> o.retryStrategy(b -> b.maxAttempts(3)))
+                .build();
+    }
+
+    @Test
+    public void testRetries()
+            throws IOException
+    {
+        S3Location location = new S3Location(Location.of("s3://bucket/object"));
+        GetObjectRequest request = GetObjectRequest.builder()
+                .bucket(location.bucket())
+                .key(location.key())
+                .build();
+        S3Client s3client = createS3Client();
+        S3Input input = new S3Input(location.location(), s3client, request);
+
+        byte[] bytes = new byte[1024];
+        assertThatThrownBy(() -> input.readFully(0, bytes, 0, 1024)).cause()
```

Comment:
*suggestion*: does it make sense to add assertion for `readTail` for completeness sake?

Reply:
Good idea, we should test that as well

Reply:
Done!